**Clean Flights 2017**
<br/>Perform cleaning on the 2017 dataframe and then export a clean csv.

In [1]:
import numpy as np
import pandas as pd

In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
# increase display size to enable viewing of all data columns

pd.options.display.max_columns = 35

In [4]:
# read in the 2017 flights csv and ensure everything loads correctly

flights2017_df = pd.read_csv('flights2017.csv', index_col=0)
flights2017_df

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
0                6  2017-07-01                AS      N559AS    DCA   
1                6  2017-07-01                AS      N513AS    SEA   
2                6  2017-07-01                AS      N588AS    DCA   
3                6  2017-07-01                AS      N538AS    SEA   
4                6  2017-07-01                AS      N536AS    DCA   
...            ...         ...               ...         ...    ...   
5674616          2  2017-09-05                UA      N456UA    EWR   
5674617          2  2017-09-05                UA      N809UA    RDU   
5674618          2  2017-09-05                UA      N69816    DEN   
5674619          2  2017-09-05                UA      N17752    EWR   
5674620          2  2017-09-05                UA      N61898    LAX   

             OriginCityName OriginState Dest           DestCityName DestState  \
0            Washington, DC          VA  SEA            Seattle, WA        WA   
1               Seattle, WA          WA  DCA         Washington, DC        VA   
2            Washington, DC          VA  SEA            Seattle, WA        WA   
3               Seattle, WA          WA  DCA         Washington, DC        VA   
4            Washington, DC          VA  LAX        Los Angeles, CA        CA   
...                     ...         ...  ...                    ...       ...   
5674616          Newark, NJ          NJ  LAS          Las Vegas, NV        NV   
5674617  Raleigh/Durham, NC          NC  DEN             Denver, CO        CO   
5674618          Denver, CO          CO  SAT        San Antonio, TX        TX   
5674619          Newark, NJ          NJ  DFW  Dallas/Fort Worth, TX        TX   
5674620     Los Angeles, CA          CA  CLE          Cleveland, OH        OH   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
0               800    750.0     -10.0       0.0                  -1.0   
1              1335   1330.0      -5.0       0.0                  -1.0   
2              1852   1856.0       4.0       0.0                   0.0   
3               800    840.0      40.0       1.0                   2.0   
4               900    854.0      -6.0       0.0                  -1.0   
...             ...      ...       ...       ...                   ...   
5674616         842    949.0      67.0       1.0                   4.0   
5674617        1623   1616.0      -7.0       0.0                  -1.0   
5674618        1526   1523.0      -3.0       0.0                  -1.0   
5674619         605    557.0      -8.0       0.0                  -1.0   
5674620        1150   1149.0      -1.0       0.0                  -1.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
0           17.0      807.0          0              NaN      2329   
1           15.0     1345.0          0              NaN      2329   
2            9.0     1905.0          0              NaN      2329   
3            9.0      849.0          0              NaN      2329   
4           13.0      907.0          0              NaN      2311   
...          ...        ...        ...              ...       ...   
5674616     24.0     1013.0          0              NaN      2227   
5674617     21.0     1637.0          0              NaN      1436   
5674618     12.0     1535.0          0              NaN       794   
5674619     24.0      621.0          0              NaN      1372   
5674620     15.0     1204.0          0              NaN      2052   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                   10           NaN           NaN       NaN            NaN   
1                   10           NaN           NaN       NaN            NaN   
2                   10           NaN           NaN       NaN            NaN   
3                   10          36.0           0.0       0.0            0.0   
4                   10           NaN           NaN       NaN          

In [5]:
# summary statistics for the dataframe

flights2017_df.describe().transpose()

count         mean         std    min     25%  \
DayOfWeek             5674621.0     3.939507    1.991768    1.0     2.0   
CRSDepTime            5674621.0  1329.968832  490.937570    1.0   912.0   
DepTime               5594313.0  1333.704391  504.398714    1.0   914.0   
DepDelay              5594278.0     9.725734   43.865896 -234.0    -5.0   
DepDel15              5594278.0     0.181229    0.385208    0.0     0.0   
DepartureDelayGroups  5594278.0     0.025027    2.149665   -2.0    -1.0   
TaxiOut               5592476.0    16.779942    9.365385    0.0    11.0   
WheelsOff             5592480.0  1355.793697  506.222749    1.0   930.0   
Cancelled             5674621.0     0.014572    0.119834    0.0     0.0   
Distance              5674621.0   856.689109  624.488107   31.0   391.0   
DistanceGroup         5674621.0     3.896428    2.456756    1.0     2.0   
CarrierDelay          1029473.0    19.929325   62.276022    0.0     0.0   
WeatherDelay          1029473.0     2.719604   24.343024    0.0     0.0   
NASDelay              1029473.0    16.028382   35.902924    0.0     0.0   
SecurityDelay         1029473.0     0.091474    3.028055    0.0     0.0   
LateAircraftDelay     1029473.0    25.163428   47.717688    0.0     0.0   
CRSArrTime            5674621.0  1489.307598  519.017512    1.0  1103.0   
ArrTime               5589947.0  1468.880705  538.398355    1.0  1050.0   
ArrDelay              5579410.0     4.326357   45.864815 -238.0   -15.0   

                         50%     75%     max  
DayOfWeek                4.0     6.0     7.0  
CRSDepTime            1323.0  1735.0  2359.0  
DepTime               1327.0  1743.0  2400.0  
DepDelay                -2.0     6.0  2755.0  
DepDel15                 0.0     0.0     1.0  
DepartureDelayGroups    -1.0     0.0    12.0  
TaxiOut                 14.0    20.0   183.0  
WheelsOff             1340.0  1758.0  2400.0  
Cancelled                0.0     0.0     1.0  
Distance               680.0  1097.0  4983.0  
DistanceGroup            3.0     5.0    11.0  
CarrierDelay             1.0    17.0  1934.0  
WeatherDelay             0.0     0.0  1934.0  
NASDelay                 2.0    19.0  1605.0  
SecurityDelay            0.0     0.0   827.0  
LateAircraftDelay        4.0    31.0  1756.0  
CRSArrTime            1520.0  1920.0  2359.0  
ArrTime               1510.0  1918.0  2400.0  
ArrDelay                -6.0     7.0  2189.0

In [6]:
# check to see if there any NaN values in the data

flights2017_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number               12511
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                   80308
DepDelay                  80343
DepDel15                  80343
DepartureDelayGroups      80343
TaxiOut                   82145
WheelsOff                 82141
Cancelled                     0
CancellationCode        5591928
Distance                      0
DistanceGroup                 0
CarrierDelay            4645148
WeatherDelay            4645148
NASDelay                4645148
SecurityDelay           4645148
LateAircraftDelay       4645148
CRSArrTime                    0
ArrTime                   84674
ArrDelay                  95211
dtype: int64

In [7]:
# list of all the columns in the data dataframe with missing values and the percent that's missing

round(np.mean(flights2017_df[flights2017_df.columns[flights2017_df.isnull().any()]].isna(), axis = 0) * 100,2)

Tail_Number              0.22
DepTime                  1.42
DepDelay                 1.42
DepDel15                 1.42
DepartureDelayGroups     1.42
TaxiOut                  1.45
WheelsOff                1.45
CancellationCode        98.54
CarrierDelay            81.86
WeatherDelay            81.86
NASDelay                81.86
SecurityDelay           81.86
LateAircraftDelay       81.86
ArrTime                  1.49
ArrDelay                 1.68
dtype: float64

In [8]:
# this shows that all flights missing the tail numbers were cancelled

flights2017_df[(flights2017_df['Cancelled'] == 1) & (flights2017_df['Tail_Number'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
34539            5  2017-07-28                OO         NaN    DHN   
76501            1  2017-07-10                OO         NaN    ORD   
76591            1  2017-07-10                OO         NaN    FWA   
76747            1  2017-07-10                OO         NaN    TVC   
77231            1  2017-07-10                OO         NaN    ORD   
...            ...         ...               ...         ...    ...   
5674594          2  2017-09-05                UA         NaN    IAH   
5674595          2  2017-09-05                UA         NaN    PDX   
5674603          2  2017-09-05                UA         NaN    IAH   
5674606          2  2017-09-05                UA         NaN    IAH   
5674610          2  2017-09-05                UA         NaN    IAH   

            OriginCityName OriginState Dest       DestCityName DestState  \
34539           Dothan, AL          AL  ATL        Atlanta, GA        GA   
76501          Chicago, IL          IL  FWA     Fort Wayne, IN        IN   
76591       Fort Wayne, IN          IN  ORD        Chicago, IL        IL   
76747    Traverse City, MI          MI  ORD        Chicago, IL        IL   
77231          Chicago, IL          IL  TVC  Traverse City, MI        MI   
...                    ...         ...  ...                ...       ...   
5674594        Houston, TX          TX  BOS         Boston, MA        MA   
5674595       Portland, OR          OR  IAH        Houston, TX        TX   
5674603        Houston, TX          TX  PHX        Phoenix, AZ        AZ   
5674606        Houston, TX          TX  LAX    Los Angeles, CA        CA   
5674610        Houston, TX          TX  BOS         Boston, MA        MA   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
34539          1059      NaN       NaN       NaN                   NaN   
76501           732      NaN       NaN       NaN                   NaN   
76591          1005      NaN       NaN       NaN                   NaN   
76747          1320      NaN       NaN       NaN                   NaN   
77231          1040      NaN       NaN       NaN                   NaN   
...             ...      ...       ...       ...                   ...   
5674594        1940      NaN       NaN       NaN                   NaN   
5674595        1230      NaN       NaN       NaN                   NaN   
5674603         909      NaN       NaN       NaN                   NaN   
5674606         910      NaN       NaN       NaN                   NaN   
5674610        1755      NaN       NaN       NaN                   NaN   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
34539        NaN        NaN          1                A       170   
76501        NaN        NaN          1                B       157   
76591        NaN        NaN          1                B       157   
76747        NaN        NaN          1                B       224   
77231        NaN        NaN          1                B       224   
...          ...        ...        ...              ...       ...   
5674594      NaN        NaN          1                C      1597   
5674595      NaN        NaN          1                C      1825   
5674603      NaN        NaN          1                C      1009   
5674606      NaN        NaN          1                C      1379   
5674610      NaN        NaN          1                C      1597   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
34539                1           NaN           NaN       NaN            NaN   
76501                1           NaN           NaN       NaN            NaN   
76591                1           NaN           NaN       NaN            NaN   
76747                1           NaN           NaN       NaN            NaN   
77231                1           NaN           NaN       NaN            NaN   
...                ...           ...           ... 

In [9]:
# fill in 'unknown' for the NaN tail numbers
# these are all flights that were cancelled, and we already have a small number of cancelled flights, so don't want to delete them

flights2017_df['Tail_Number'].fillna('unknown', inplace=True)
flights2017_df.shape

(5674621, 29)

In [10]:
# check to ensure that tail number was filled in appropriately

flights2017_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                   80308
DepDelay                  80343
DepDel15                  80343
DepartureDelayGroups      80343
TaxiOut                   82145
WheelsOff                 82141
Cancelled                     0
CancellationCode        5591928
Distance                      0
DistanceGroup                 0
CarrierDelay            4645148
WeatherDelay            4645148
NASDelay                4645148
SecurityDelay           4645148
LateAircraftDelay       4645148
CRSArrTime                    0
ArrTime                   84674
ArrDelay                  95211
dtype: int64

In [11]:
# check to see how many cancelled flights there were in 2017

flights2017_df[flights2017_df['Cancelled'] == 1].shape

(82693, 29)

82,693 cancelled flights, but 80,308 with no departure time -- there are 2,385 (see below) that need updating.

In [12]:
# look at the flights that are both cancelled and have a departure time

flights2017_df[(flights2017_df['Cancelled'] ==1) & (flights2017_df['DepTime'])]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
3604             5  2017-07-07                AS      N526AS    LAX   
4618             7  2017-07-09                AS      N320AS    BNA   
4936             1  2017-07-10                AS      N618AS    ANC   
4957             1  2017-07-10                AS      N524AS    SIT   
5368             1  2017-07-10                AS      N407AS    BWI   
...            ...         ...               ...         ...    ...   
5651439          1  2017-09-11                UA      N818UA    ORD   
5662110          5  2017-09-08                UA      N77431    IAD   
5670557          3  2017-09-20                UA      N17244    SFO   
5671096          2  2017-09-19                UA      N78506    BZN   
5672259          2  2017-09-19                UA      N57869    ORD   

            OriginCityName OriginState Dest       DestCityName DestState  \
3604       Los Angeles, CA          CA  PDX       Portland, OR        OR   
4618         Nashville, TN          TN  SEA        Seattle, WA        WA   
4936         Anchorage, AK          AK  BET         Bethel, AK        AK   
4957             Sitka, AK          AK  SEA        Seattle, WA        WA   
5368         Baltimore, MD          MD  SEA        Seattle, WA        WA   
...                    ...         ...  ...                ...       ...   
5651439        Chicago, IL          IL  CHS     Charleston, SC        SC   
5662110     Washington, DC          VA  SFO  San Francisco, CA        CA   
5670557  San Francisco, CA          CA  DEN         Denver, CO        CO   
5671096        Bozeman, MT          MT  DEN         Denver, CO        CO   
5672259        Chicago, IL          IL  SJU       San Juan, PR        PR   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
3604            830    832.0       2.0       0.0                   0.0   
4618           1510   1502.0      -8.0       0.0                  -1.0   
4936            620    609.0     -11.0       0.0                  -1.0   
4957           1357   1704.0     187.0       1.0                  12.0   
5368           1840   1851.0      11.0       0.0                   0.0   
...             ...      ...       ...       ...                   ...   
5651439         734    737.0       3.0       0.0                   0.0   
5662110        1935   1925.0     -10.0       0.0                  -1.0   
5670557        2030   2141.0      71.0       1.0                   4.0   
5671096         714    718.0       4.0       0.0                   0.0   
5672259         750    819.0      29.0       1.0                   1.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
3604         NaN        NaN          1                A       834   
4618         NaN        NaN          1                A      1978   
4936        12.0      621.0          1                B       399   
4957         NaN        NaN          1                A       861   
5368         NaN        NaN          1                A      2335   
...          ...        ...        ...              ...       ...   
5651439     20.0      757.0          1                B       760   
5662110      NaN        NaN          1                A      2419   
5670557      NaN        NaN          1                A       967   
5671096      NaN        NaN          1                A       524   
5672259     23.0      842.0          1                A      2072   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
3604                 4           NaN           NaN       NaN            NaN   
4618                 8           NaN           NaN       NaN            NaN   
4936                 2           NaN           NaN       NaN            NaN   
4957                 4           NaN           NaN       NaN            NaN   
5368                10           NaN           NaN       NaN            NaN   
...                ...           ...           ... 

These don't have arrival times, so it is safe to assume that they were actually cancelled, and should not have a departure time.

In [13]:
# change the departure time to 0 for all flights that are cancelled and have a depature time
# and check that there are now no flights with this combination

flights2017_df['DepTime'][(flights2017_df['Cancelled'] == 1) & (flights2017_df['DepTime'])] = 0
flights2017_df[(flights2017_df['Cancelled'] ==1) & (flights2017_df['DepTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [14]:
# check for flights that have a depature time greater than 0, but no departure delay

flights2017_df[(flights2017_df['DepTime'] > 0 ) & (flights2017_df['DepDelay'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
2159957          7  2017-10-29                F9      N713FR    MCO   

        OriginCityName OriginState Dest    DestCityName DestState  CRSDepTime  \
2159957    Orlando, FL          FL  PVD  Providence, RI        RI        1915   

         DepTime  DepDelay  DepDel15  DepartureDelayGroups  TaxiOut  \
2159957   1925.0       NaN       NaN                   NaN    175.0   

         WheelsOff  Cancelled CancellationCode  Distance  DistanceGroup  \
2159957     2220.0          0              NaN      1072              5   

         CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
2159957           NaN           NaN       NaN            NaN   

         LateAircraftDelay  CRSArrTime  ArrTime  ArrDelay  
2159957                NaN        2205      NaN       NaN

In [15]:
# there is only one flight in this situation, so manually fill in the appropriate values for DepDelay, DepDel15, and DepatureDelayGroups

flights2017_df['DepDelay'][(flights2017_df['DepTime'] > 0 ) & (flights2017_df['DepDelay'].isna())] = 10
flights2017_df['DepDel15'][(flights2017_df['DepTime'] > 0 ) & (flights2017_df['DepDel15'].isna())] = 0
flights2017_df['DepartureDelayGroups'][(flights2017_df['DepTime'] > 0 ) & (flights2017_df['DepartureDelayGroups'].isna())] = 0

In [16]:
# review remaining flights with NaN values in the DepDelay column

flights2017_df[(flights2017_df['DepDelay'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
581              7  2017-07-02                AS      N765AS    BRW   
1157             1  2017-07-03                AS      N763AS    PSG   
1389             1  2017-07-03                AS      N615AS    SEA   
1394             1  2017-07-03                AS      N644AS    SAN   
1490             1  2017-07-03                AS      N546AS    DFW   
...            ...         ...               ...         ...    ...   
5674594          2  2017-09-05                UA     unknown    IAH   
5674595          2  2017-09-05                UA     unknown    PDX   
5674603          2  2017-09-05                UA     unknown    IAH   
5674606          2  2017-09-05                UA     unknown    IAH   
5674610          2  2017-09-05                UA     unknown    IAH   

                OriginCityName OriginState Dest     DestCityName DestState  \
581                 Barrow, AK          AK  FAI    Fairbanks, AK        AK   
1157            Petersburg, AK          AK  WRG     Wrangell, AK        AK   
1389               Seattle, WA          WA  SAN    San Diego, CA        CA   
1394             San Diego, CA          CA  SEA      Seattle, WA        WA   
1490     Dallas/Fort Worth, TX          TX  SEA      Seattle, WA        WA   
...                        ...         ...  ...              ...       ...   
5674594            Houston, TX          TX  BOS       Boston, MA        MA   
5674595           Portland, OR          OR  IAH      Houston, TX        TX   
5674603            Houston, TX          TX  PHX      Phoenix, AZ        AZ   
5674606            Houston, TX          TX  LAX  Los Angeles, CA        CA   
5674610            Houston, TX          TX  BOS       Boston, MA        MA   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
581            1150      NaN       NaN       NaN                   NaN   
1157           1509      NaN       NaN       NaN                   NaN   
1389           1355      NaN       NaN       NaN                   NaN   
1394            800      NaN       NaN       NaN                   NaN   
1490           1920      NaN       NaN       NaN                   NaN   
...             ...      ...       ...       ...                   ...   
5674594        1940      NaN       NaN       NaN                   NaN   
5674595        1230      NaN       NaN       NaN                   NaN   
5674603         909      NaN       NaN       NaN                   NaN   
5674606         910      NaN       NaN       NaN                   NaN   
5674610        1755      NaN       NaN       NaN                   NaN   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
581          NaN        NaN          1                B       503   
1157         NaN        NaN          1                B        31   
1389         NaN        NaN          1                A      1050   
1394         NaN        NaN          1                A      1050   
1490         NaN        NaN          1                A      1660   
...          ...        ...        ...              ...       ...   
5674594      NaN        NaN          1                C      1597   
5674595      NaN        NaN          1                C      1825   
5674603      NaN        NaN          1                C      1009   
5674606      NaN        NaN          1                C      1379   
5674610      NaN        NaN          1                C      1597   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
581                  3           NaN           NaN       NaN            NaN   
1157                 1           NaN           NaN       NaN            NaN   
1389                 5           NaN           NaN       NaN            NaN   
1394                 5           NaN           NaN       NaN            NaN   
1490                 7           NaN           NaN       NaN            NaN   
...                ...     

In [17]:
# fill in 0 for the remaining NaN values in the DepTime, DepDelay, DepDel15 and DepartureDelayGroups columns

flights2017_df['DepTime'].fillna(0, inplace=True)
flights2017_df['DepDelay'].fillna(0, inplace=True)
flights2017_df['DepDel15'].fillna(0, inplace=True)
flights2017_df['DepartureDelayGroups'].fillna(0, inplace=True)
flights2017_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                   82145
WheelsOff                 82141
Cancelled                     0
CancellationCode        5591928
Distance                      0
DistanceGroup                 0
CarrierDelay            4645148
WeatherDelay            4645148
NASDelay                4645148
SecurityDelay           4645148
LateAircraftDelay       4645148
CRSArrTime                    0
ArrTime                   84674
ArrDelay                  95211
dtype: int64

In [18]:
# look at the flights that are both cancelled and have a departure time

flights2017_df[(flights2017_df['Cancelled'] ==1) & (flights2017_df['ArrTime'])]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
4216847          1  2017-03-06                NK      N509NK    OAK   
4216896          1  2017-03-13                NK      N588NK    FLL   
4216927          1  2017-03-06                NK      N615NK    DEN   
4216931          1  2017-03-06                NK      N615NK    LAS   
4216935          1  2017-03-06                NK      N615NK    LAS   
...            ...         ...               ...         ...    ...   
4695729          7  2017-03-12                AA      N4UBAA    DEN   
4695733          7  2017-03-12                AA      N566AA    ORD   
4696598          7  2017-03-12                AA      N3HGAA    ABQ   
4696672          7  2017-03-12                AA      N4XMAA    ORD   
4700145          7  2017-03-05                WN      N463WN    LAS   

              OriginCityName OriginState Dest           DestCityName  \
4216847          Oakland, CA          CA  LAX        Los Angeles, CA   
4216896  Fort Lauderdale, FL          FL  EWR             Newark, NJ   
4216927           Denver, CO          CO  LAS          Las Vegas, NV   
4216931        Las Vegas, NV          NV  LAX        Los Angeles, CA   
4216935        Las Vegas, NV          NV  DEN             Denver, CO   
...                      ...         ...  ...                    ...   
4695729           Denver, CO          CO  ORD            Chicago, IL   
4695733          Chicago, IL          IL  DEN             Denver, CO   
4696598      Albuquerque, NM          NM  DFW  Dallas/Fort Worth, TX   
4696672          Chicago, IL          IL  DFW  Dallas/Fort Worth, TX   
4700145        Las Vegas, NV          NV  OAK            Oakland, CA   

        DestState  CRSDepTime  DepTime  DepDelay  DepDel15  \
4216847        CA        2115      0.0       0.0       0.0   
4216896        NJ        2050      0.0       0.0       0.0   
4216927        NV        1131      0.0       0.0       0.0   
4216931        CA        1313      0.0       0.0       0.0   
4216935        CO         755      0.0       0.0       0.0   
...           ...         ...      ...       ...       ...   
4695729        IL        1417      0.0       0.0       0.0   
4695733        CO        1200      0.0       0.0       0.0   
4696598        TX        1500      0.0       0.0       0.0   
4696672        TX        1025      0.0       0.0       0.0   
4700145        CA        2150      0.0       0.0       0.0   

         DepartureDelayGroups  TaxiOut  WheelsOff  Cancelled CancellationCode  \
4216847                   0.0      NaN        NaN          1                C   
4216896                   0.0      NaN        NaN          1                B   
4216927                   0.0      NaN        NaN          1                C   
4216931                   0.0      NaN        NaN          1                C   
4216935                   0.0      NaN        NaN          1                C   
...                       ...      ...        ...        ...              ...   
4695729                   0.0      NaN        NaN          1                A   
4695733                   0.0      NaN        NaN          1                A   
4696598                   0.0      NaN        NaN          1                A   
4696672                   0.0      NaN        NaN          1                A   
4700145                   0.0      NaN        NaN          1                C   

         Distance  DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  \
4216847       337              2           NaN           NaN       NaN   
4216896      1065              5           NaN           NaN       NaN   
4216927       628              3           NaN           NaN       NaN   
4216931       236              1           NaN           NaN       NaN   
4216935       628              3           NaN           NaN       NaN   
...           ...            ...           ...           ...       ...   
4695729       888              4           NaN         

In [19]:
# change the arrival time and arrival delay to 0 for all flights that are cancelled and have an arrival time
# and check that there are now no flights with this combination

flights2017_df['ArrTime'][(flights2017_df['Cancelled'] == 1) & (flights2017_df['ArrTime'])] = 0
flights2017_df['ArrDelay'][(flights2017_df['Cancelled'] == 1) & (flights2017_df['ArrDelay'])] = 0
flights2017_df[(flights2017_df['Cancelled'] == 1) & (flights2017_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [20]:
# inspect the columns with NaN values for arrival time that are not cancelled

flights2017_df[(flights2017_df['Cancelled'] == 0) & (flights2017_df['ArrTime'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
583              7  2017-07-02                AS      N765AS    SCC   
1156             1  2017-07-03                AS      N763AS    JNU   
2513             3  2017-07-05                AS      N318AS    SEA   
3061             4  2017-07-06                AS      N529AS    PDX   
3732             5  2017-07-07                AS      N563AS    DCA   
...            ...         ...               ...         ...    ...   
5650422          1  2017-09-11                UA      N17229    SFO   
5670034          3  2017-09-20                UA      N76529    IAH   
5671890          2  2017-09-19                UA      N14250    SFO   
5673914          2  2017-09-05                UA      N54711    EWR   
5674319          2  2017-09-05                UA      N16713    SNA   

            OriginCityName OriginState Dest    DestCityName DestState  \
583          Deadhorse, AK          AK  BRW      Barrow, AK        AK   
1156            Juneau, AK          AK  PSG  Petersburg, AK        AK   
2513           Seattle, WA          WA  LAS   Las Vegas, NV        NV   
3061          Portland, OR          OR  SNA   Santa Ana, CA        CA   
3732        Washington, DC          VA  PDX    Portland, OR        OR   
...                    ...         ...  ...             ...       ...   
5650422  San Francisco, CA          CA  SNA   Santa Ana, CA        CA   
5670034        Houston, TX          TX  SNA   Santa Ana, CA        CA   
5671890  San Francisco, CA          CA  BUR     Burbank, CA        CA   
5673914         Newark, NJ          NJ  SNA   Santa Ana, CA        CA   
5674319      Santa Ana, CA          CA  ORD     Chicago, IL        IL   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
583            1006   1005.0      -1.0       0.0                  -1.0   
1156           1341   1406.0      25.0       1.0                   1.0   
2513           2150   2145.0      -5.0       0.0                  -1.0   
3061           1540   1903.0     203.0       1.0                  12.0   
3732           1720   1710.0     -10.0       0.0                  -1.0   
...             ...      ...       ...       ...                   ...   
5650422        2102   2221.0      79.0       1.0                   5.0   
5670034        2108   2107.0      -1.0       0.0                  -1.0   
5671890        2015   2009.0      -6.0       0.0                  -1.0   
5673914        1833   2122.0     169.0       1.0                  11.0   
5674319         655    647.0      -8.0       0.0                  -1.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
583          6.0     1011.0          0              NaN       204   
1156        19.0     1425.0          0              NaN       123   
2513        20.0     2205.0          0              NaN       867   
3061        21.0     1924.0          0              NaN       859   
3732        28.0     1738.0          0              NaN      2350   
...          ...        ...        ...              ...       ...   
5650422     21.0     2242.0          0              NaN       372   
5670034     14.0     2121.0          0              NaN      1346   
5671890     16.0     2025.0          0              NaN       326   
5673914     75.0     2237.0          0              NaN      2434   
5674319     18.0      705.0          0              NaN      1726   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
583                  1           NaN           NaN       NaN            NaN   
1156                 1           NaN           NaN       NaN            NaN   
2513                 4           NaN           NaN       NaN            NaN   
3061                 4           NaN           NaN       NaN            NaN   
3732                10           NaN           NaN       NaN            NaN   
...                ...           ...           ...       ...            ...   
5650422 

In [21]:
# there are 10,318 rows with flights that are not cancelled, and have a departure time, but no arrival time
# delete these rows

flights2017_df = flights2017_df.drop(flights2017_df[(flights2017_df['Cancelled'] == 0) & (flights2017_df['ArrTime'].isna())].index)
flights2017_df[(flights2017_df['Cancelled'] == 0) & (flights2017_df['ArrTime'].isna())]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [22]:
# replace NaN with 0 for all flights that are cancelled 
# and check that there are now no flights with this combination

flights2017_df['ArrTime'][(flights2017_df['Cancelled'] == 1) & (flights2017_df['ArrTime'].isna())] = 0
flights2017_df['ArrDelay'][(flights2017_df['Cancelled'] == 1) & (flights2017_df['ArrDelay'].isna())] = 0
flights2017_df[(flights2017_df['Cancelled'] == 1) & (flights2017_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [23]:
# check the status of NaNs

flights2017_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                   82145
WheelsOff                 82141
Cancelled                     0
CancellationCode        5581610
Distance                      0
DistanceGroup                 0
CarrierDelay            4636458
WeatherDelay            4636458
NASDelay                4636458
SecurityDelay           4636458
LateAircraftDelay       4636458
CRSArrTime                    0
ArrTime                       0
ArrDelay                  10531
dtype: int64

In [24]:
# look at the remaining flights with NaN values for ArrDelay

flights2017_df[(flights2017_df['ArrDelay'].isna()) & (flights2017_df['ArrTime'])]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
452              6  2017-07-01                AS      N247AK    PDX   
1133             1  2017-07-03                AS      N577AS    ANC   
2550             3  2017-07-05                AS      N481AS    SEA   
4740             7  2017-07-09                AS      N585AS    SEA   
4900             1  2017-07-10                AS      N525AS    LAX   
...            ...         ...               ...         ...    ...   
5673978          2  2017-09-05                UA      N852UA    ORD   
5673987          2  2017-09-05                UA      N63890    SAN   
5674252          2  2017-09-05                UA      N69819    LAS   
5674261          2  2017-09-05                UA      N24729    PHX   
5674347          2  2017-09-05                UA      N596UA    LAX   

          OriginCityName OriginState Dest           DestCityName DestState  \
452         Portland, OR          OR  MCO            Orlando, FL        FL   
1133       Anchorage, AK          AK  SCC          Deadhorse, AK        AK   
2550         Seattle, WA          WA  DFW  Dallas/Fort Worth, TX        TX   
4740         Seattle, WA          WA  DFW  Dallas/Fort Worth, TX        TX   
4900     Los Angeles, CA          CA  DCA         Washington, DC        VA   
...                  ...         ...  ...                    ...       ...   
5673978      Chicago, IL          IL  LGA           New York, NY        NY   
5673987    San Diego, CA          CA  EWR             Newark, NJ        NJ   
5674252    Las Vegas, NV          NV  EWR             Newark, NJ        NJ   
5674261      Phoenix, AZ          AZ  EWR             Newark, NJ        NJ   
5674347  Los Angeles, CA          CA  EWR             Newark, NJ        NJ   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
452             640    639.0      -1.0       0.0                  -1.0   
1133            737    727.0     -10.0       0.0                  -1.0   
2550           1220   1212.0      -8.0       0.0                  -1.0   
4740            700    654.0      -6.0       0.0                  -1.0   
4900           1230   1250.0      20.0       1.0                   1.0   
...             ...      ...       ...       ...                   ...   
5673978        1500   1458.0      -2.0       0.0                  -1.0   
5673987        1125   1201.0      36.0       1.0                   2.0   
5674252        1120   1118.0      -2.0       0.0                  -1.0   
5674261        1036   1218.0     102.0       1.0                   6.0   
5674347         940    934.0      -6.0       0.0                  -1.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
452         18.0      657.0          0              NaN      2534   
1133         8.0      735.0          0              NaN       627   
2550        24.0     1236.0          0              NaN      1660   
4740        15.0      709.0          0              NaN      1660   
4900        12.0     1302.0          0              NaN      2311   
...          ...        ...        ...              ...       ...   
5673978     30.0     1528.0          0              NaN       733   
5673987     20.0     1221.0          0              NaN      2425   
5674252     31.0     1149.0          0              NaN      2227   
5674261     19.0     1237.0          0              NaN      2133   
5674347     20.0      954.0          0              NaN      2454   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
452                 11           NaN           NaN       NaN            NaN   
1133                 3           NaN           NaN       NaN            NaN   
2550                 7           NaN           NaN       NaN            NaN   
4740                 7           NaN           NaN       NaN            NaN   
4900                10           NaN           NaN       NaN            NaN   
...                ...     

In [25]:
# fill in these NaN values with the difference between the expected arrival time (CRSArrTime) and the actual arrival time (ArrTime)

flights2017_df['ArrDelay'].fillna((flights2017_df['ArrTime'] - flights2017_df['CRSArrTime']), inplace=True)
flights2017_df[(flights2017_df['ArrDelay'].isna()) & (flights2017_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [26]:
# check the status of NaNs

flights2017_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                   82145
WheelsOff                 82141
Cancelled                     0
CancellationCode        5581610
Distance                      0
DistanceGroup                 0
CarrierDelay            4636458
WeatherDelay            4636458
NASDelay                4636458
SecurityDelay           4636458
LateAircraftDelay       4636458
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [27]:
# fill in the TaxiOut column with the median value 

flights2017_df['TaxiOut'].fillna(flights2017_df['TaxiOut'].median(), inplace=True)
flights2017_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                       0
WheelsOff                 82141
Cancelled                     0
CancellationCode        5581610
Distance                      0
DistanceGroup                 0
CarrierDelay            4636458
WeatherDelay            4636458
NASDelay                4636458
SecurityDelay           4636458
LateAircraftDelay       4636458
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [28]:
# fill in WheelsOff with the CRSDepTime + the TaxiOut time

flights2017_df['WheelsOff'].fillna((flights2017_df['CRSDepTime'] + flights2017_df['TaxiOut']), inplace=True)
flights2017_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                       0
WheelsOff                     0
Cancelled                     0
CancellationCode        5581610
Distance                      0
DistanceGroup                 0
CarrierDelay            4636458
WeatherDelay            4636458
NASDelay                4636458
SecurityDelay           4636458
LateAircraftDelay       4636458
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [29]:
# check the shapes of the whole dataframe and just the cancelled flights

display(flights2017_df.shape)
display(flights2017_df[flights2017_df['Cancelled'] == 1].shape)

(5664303, 29)

(82693, 29)

82,693 cancelled + 5,581,610 NaN Cancelation Code = 5,664,303 total

In [30]:
# replace NaN with 0 for CancellationCode and the reason for Delay

flights2017_df['CancellationCode'].fillna(0, inplace=True)
flights2017_df['CarrierDelay'].fillna(0, inplace=True)
flights2017_df['WeatherDelay'].fillna(0, inplace=True)
flights2017_df['NASDelay'].fillna(0, inplace=True)
flights2017_df['SecurityDelay'].fillna(0, inplace=True)
flights2017_df['LateAircraftDelay'].fillna(0, inplace=True)
flights2017_df.isna().sum()

DayOfWeek               0
FlightDate              0
Reporting_Airline       0
Tail_Number             0
Origin                  0
OriginCityName          0
OriginState             0
Dest                    0
DestCityName            0
DestState               0
CRSDepTime              0
DepTime                 0
DepDelay                0
DepDel15                0
DepartureDelayGroups    0
TaxiOut                 0
WheelsOff               0
Cancelled               0
CancellationCode        0
Distance                0
DistanceGroup           0
CarrierDelay            0
WeatherDelay            0
NASDelay                0
SecurityDelay           0
LateAircraftDelay       0
CRSArrTime              0
ArrTime                 0
ArrDelay                0
dtype: int64

In [31]:
# add a boolean column for whether a flight is delayed or not

flights2017_df['Delayed'] = 0

flights2017_df.loc[flights2017_df['ArrDelay'] <= 0, 'Delayed'] = 0
flights2017_df.loc[flights2017_df['ArrDelay'] > 0, 'Delayed'] = 1

flights2017_df['Delayed'].value_counts()

0    3713977
1    1950326
Name: Delayed, dtype: int64

In [32]:
# display the clean dataframe

flights2017_df

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
0                6  2017-07-01                AS      N559AS    DCA   
1                6  2017-07-01                AS      N513AS    SEA   
2                6  2017-07-01                AS      N588AS    DCA   
3                6  2017-07-01                AS      N538AS    SEA   
4                6  2017-07-01                AS      N536AS    DCA   
...            ...         ...               ...         ...    ...   
5674616          2  2017-09-05                UA      N456UA    EWR   
5674617          2  2017-09-05                UA      N809UA    RDU   
5674618          2  2017-09-05                UA      N69816    DEN   
5674619          2  2017-09-05                UA      N17752    EWR   
5674620          2  2017-09-05                UA      N61898    LAX   

             OriginCityName OriginState Dest           DestCityName DestState  \
0            Washington, DC          VA  SEA            Seattle, WA        WA   
1               Seattle, WA          WA  DCA         Washington, DC        VA   
2            Washington, DC          VA  SEA            Seattle, WA        WA   
3               Seattle, WA          WA  DCA         Washington, DC        VA   
4            Washington, DC          VA  LAX        Los Angeles, CA        CA   
...                     ...         ...  ...                    ...       ...   
5674616          Newark, NJ          NJ  LAS          Las Vegas, NV        NV   
5674617  Raleigh/Durham, NC          NC  DEN             Denver, CO        CO   
5674618          Denver, CO          CO  SAT        San Antonio, TX        TX   
5674619          Newark, NJ          NJ  DFW  Dallas/Fort Worth, TX        TX   
5674620     Los Angeles, CA          CA  CLE          Cleveland, OH        OH   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
0               800    750.0     -10.0       0.0                  -1.0   
1              1335   1330.0      -5.0       0.0                  -1.0   
2              1852   1856.0       4.0       0.0                   0.0   
3               800    840.0      40.0       1.0                   2.0   
4               900    854.0      -6.0       0.0                  -1.0   
...             ...      ...       ...       ...                   ...   
5674616         842    949.0      67.0       1.0                   4.0   
5674617        1623   1616.0      -7.0       0.0                  -1.0   
5674618        1526   1523.0      -3.0       0.0                  -1.0   
5674619         605    557.0      -8.0       0.0                  -1.0   
5674620        1150   1149.0      -1.0       0.0                  -1.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
0           17.0      807.0          0                0      2329   
1           15.0     1345.0          0                0      2329   
2            9.0     1905.0          0                0      2329   
3            9.0      849.0          0                0      2329   
4           13.0      907.0          0                0      2311   
...          ...        ...        ...              ...       ...   
5674616     24.0     1013.0          0                0      2227   
5674617     21.0     1637.0          0                0      1436   
5674618     12.0     1535.0          0                0       794   
5674619     24.0      621.0          0                0      1372   
5674620     15.0     1204.0          0                0      2052   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                   10           0.0           0.0       0.0            0.0   
1                   10           0.0           0.0       0.0            0.0   
2                   10           0.0           0.0       0.0            0.0   
3                   10          36.0           0.0       0.0            0.0   
4                   10           0.0           0.0       0.0          

In [33]:
# export the clean dataframe with no NaN values to a csv

flights2017_df.to_csv('flights2017clean.csv')